### PR TITLE
Confirmation button

### PR DESCRIPTION
**Pull request by osdiab**
Shows a confirmation modal before acting.

### CODE REVIEW

File: src/ConfirmationButton/ConfirmationButton.jsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  toggleConfirmation(show) {
+    this.setState({showConfirmation: show});
+  }
+
+  render() {
+    const modalContents = (<div>
+      <p>{this.props.confirmationDescription}</p>
+      <p>
+        <Button type="secondary" value="Cancel"
+          onClick={_.partial(this.toggleConfirmation, false)}
```
Comment (jacobsimon) [inline src/ConfirmationButton/ConfirmationButton.jsx]: The partial application syntax is kind of confusing tbh, how do you feel about:
* es6 syntax: `onClick={() => this.toggleConfirmation(false)}` (this is what they do in the react docs)
* explicit handlers like `showConfirmation` and `hideConfirmation`
* just actually toggle state of taking an arg `setState({show: !this.state.show})`

File: src/ConfirmationButton/ConfirmationButton.jsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        />
+        <Button type={this.props.confirmButtonType} value={this.props.confirmButtonText}
+          onClick={_.compose(_.partial(this.toggleConfirmation, false), this.props.onConfirm)}
```
Comment (jacobsimon) [inline src/ConfirmationButton/ConfirmationButton.jsx]: see comment above

File: src/ConfirmationButton/ConfirmationButton.jsx [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        />
+      </p>
+    </div>);
+
+    // don't pass properties not for buttons, nor onClick, to the underlying Button to display.
+    const buttonProps = _.omit(_.pick(this.props, _.keys(Button.propTypes)), "onClick");
+
+    return (<div className="ConfirmationButton">
+      <Button {...buttonProps} onClick={_.partial(this.toggleConfirmation, true)} />
+      {this.state.showConfirmation ? <Modal title="Confirm regenerating badges"
```
Comment (jacobsimon) [inline src/ConfirmationButton/ConfirmationButton.jsx]: this should be `title={this.props.confirmationHeader}`

File: src/ConfirmationButton/ConfirmationButton.jsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import React, {Component, PropTypes} from "react";
+import _ from "underscore";
+
+import {Button} from "../Button/Button";
```
Comment (jacobsimon) [inline src/ConfirmationButton/ConfirmationButton.jsx]: haven't tried this but can you just do `import {Button} from ".."`

**Comment by jacobsimon** [comment]
Two thoughts: 
* There are cases where you want to have a confirmation modal, but you might not want/need a Button to trigger it (you're leaving the page, you're clicking a link or dropdown option, etc.). Maybe it would be better to create a ConfirmationDialog component and have ConfirmationButton consume that? 
* We should double check what happens when the client is using a different version of underscore, and use the webpack de-dupe plugin

File: src/ConfirmationButton/ConfirmationButton.jsx (hunk 1 above)
Comment (osdiab) [inline src/ConfirmationButton/ConfirmationButton.jsx]: 929b690

File: src/ConfirmationButton/ConfirmationButton.jsx (hunk 2 above)
Comment (osdiab) [inline src/ConfirmationButton/ConfirmationButton.jsx]: 929b690

File: src/ConfirmationButton/ConfirmationButton.jsx (hunk 3 above)
Comment (osdiab) [inline src/ConfirmationButton/ConfirmationButton.jsx]: 4b8399d

**Comment by osdiab** [comment]
With respect to underscore, there exist ES6 features that do the same thing; if we add in the `Object.assign` polyfill since node doesn't yet support that, then we don't need it for `_.extend`; and I can replace `_.omit` and `_.pick` by using `reduce`.

File: src/ConfirmationButton/ConfirmationButton.jsx (hunk 4 above)
Comment (osdiab) [inline src/ConfirmationButton/ConfirmationButton.jsx]: ac666a6

**Comment by osdiab** [comment]
I will modify this to be two components: `ModalButton` and `ConfirmationButton`.

`ModalButton` will be a button that opens and closes a modal. `ConfirmationButton` will be a special instance of `ModalButton` that presents a dialog for confirmation.

**Comment by osdiab** [comment]
Done! I modified the approach so that rather than trying to rename the properties on Button and Modal, it just subsumes them; if this is a more confusing API I can do the propType omission/readdition tricks I did before.

Not too shabby!

**Comment by osdiab** [comment]
Errors about bad `propTypes` are fixed in #22

**Comment by jacobsimon** [comment]
I like it! And agree about closeModal/onClose. It would be really cool later on to generalize the modal handling logic as a higher order component that could apply to any component (custom or native) that you want to trigger a modal, e.g. `var ModalButton = ModalHandler(Button)`

**Comment by osdiab** [comment]
That would be cool—I think we'd need some kind of standardized hook to do that composition, which is something we can think about next time this sort of behavior comes up.

**Comment by osdiab** [comment]
I think another potentially valuable generalization would be to allow for more than one "action" button in the `ConfirmationButton`. Do you think that's valuable as part of this component, or can be made a more general one with this as a special, common case?

**Comment by osdiab** [comment]
Check this latest edit @jacobsimon; I allow for the full breadth of `Button` options in `ConfirmationButton` by automatically prefixing `Button`'s props with `"confirmButton"`; and prefix `Modal` options in `ModalButton` with `"modal"`. Is this useful/does this make sense? I think it does!

**Comment by jacobsimon** [comment]
LGTM!

**Comment by osdiab** [comment]
@jacobsimon lookover the test style and the stylesheets pl0x! If good then i'll merge.

File: package.json [hunk 5]
```diff
@@ -36,6 +36,7 @@
     "less": "^2.6.1",
     "less-loader": "^2.2.3",
     "mocha": "^2.4.5",
+    "object-assign": "^4.0.1",
```
Comment (jacobsimon) [inline package.json]: do we still need this since we have the babel plugin?

File: test/ConfirmationButton_test.jsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import React from "react";
+import sinon from "sinon";
+import {mount, shallow} from "enzyme";
+import {Button, ModalButton, ConfirmationButton} from "../src";
+
+describe("ConfirmationButton", () => {
+  it("renders a modal with confirm and cancel buttons", () => {
+    const confirmationButton = shallow(<ConfirmationButton value="Confirm pl0x" modalTitle="A title">
+      <p>Hello</p>
+    </ConfirmationButton>);
```
Comment (jacobsimon) [inline test/ConfirmationButton_test.jsx]: nit: we've been doing indentation in other tests like this https://github.com/Clever/components/blob/master/test/Modal_test.jsx#L9-L13
```
shallow(
  <ConfirmationButton> 
   <p>Hello</p>
  </ConfirmationButton>
); 
```

File: test/ConfirmationButton_test.jsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    assert.equal(confirmationButton.find(ModalButton).length, 1);
+    assert(!!confirmationButton.find(ModalButton).props().children.find(
+      (child) => {
```
Comment (jacobsimon) [inline test/ConfirmationButton_test.jsx]: could we express this test with selectors like you do in later tests? for instance you could do something like
```js
confirmationButton.find("button").simulate("click");
const cancelButton = confirmationButton.find("button.ConfirmationButton--dialog-buttons[value='Cancel'])");
const confirmButton = confirmationButton.find("button.ConfirmationButton--dialog-buttons[value='Confirm'])");
assert(cancelButton.length, 1);
assert(confirmButton.length, 1);
```

Comment (osdiab) [inline test/ConfirmationButton_test.jsx]: 0856ed6

File: test/ConfirmationButton_test.jsx (hunk 6 above)
Comment (osdiab) [inline test/ConfirmationButton_test.jsx]: 0856ed6

File: package.json (hunk 5 above)
Comment (osdiab) [inline package.json]: 0856ed6

**Comment by osdiab** [comment]
@jacobsimon last look?

**Comment by jacobsimon** [comment]
looks good!

**Pull request merged**